src/Apropos/Gen.hs

```diff
@@ -185,35 +226,40 @@ gen (Pure a) = pure a
     Right x -> gen $ b x
     Left e -> throwE e
 
-(>>>=) :: Generator r -> (r -> Gen a) -> Generator a
-(>>>=) a b = do
+(>=>=) :: Generator r -> (r -> Gen a) -> Generator a
+(>=>=) a b = do
   r <- lift (runExceptT a)
   case r of
     Right x -> gen $ b x
     Left e -> throwE e
 
-handleRootRetries :: Int -> Generator a -> GenContext a
-handleRootRetries 0 _ = H.footnote "global retry limit reached reached" >> H.failure
-handleRootRetries l g = do
-  gr <- runExceptT g
-  case gr of
-    Right a -> pure a
-    Left Retry -> handleRootRetries (l - 1) g
-    Left (GenException e) -> H.footnote e >> H.failure
+fromPred :: (a -> Bool) -> a -> Maybe a
+fromPred p a = a <$ guard (p a)
 
-genFilter' :: forall r. (r -> Bool) -> Gen r -> Generator r
-genFilter' condition g = go 100
-  where
-    go :: Int -> Generator r
-    go l = do
-      if l < 0
-        then throwE Retry
-        else do
-          res <- gen g
-          if condition res
-            then pure res
-            else go (l - 1)
+filter' :: (a -> Bool) -> Generator a -> Generator a
+filter' p =
+  mapMaybe (fromPred p)
+
+mapMaybe :: forall a b. (a -> Maybe b) -> Generator a -> Generator b
+mapMaybe p gen0 =
+  let try k =
+        if k > 100
+          then throwE Retry
+          else do
+            (x, g) <- lift $ lift $ HGen.freeze $ HGen.scale (2 * k +) (runWriterT (runExceptT gen0))
+            case (p <$>) $ fst x of
+              Right (Just _) -> do
+                let withGenT f = H.fromGenT . f . H.toGenT
+                let toMaybe' z = case fst z of
+                      Right a -> a
+                      Left _ -> error "This should be impossible."
+                lift $ tell $ snd x
+                lift $ lift $ withGenT (HIG.mapGenT (HIT.mapMaybeMaybeT p)) (toMaybe' <$> g)
+              Left (GenException e) -> throwE $ GenException e
+              _ -> try (k + 1)
+   in try 0
```

Comment:
This takes the `HedgeHog` definition of `filter` and makes it work with retries. If the `Gen` being filtered throws a retry this just increments the counter in filter. If the filter runs out of retries we throw a `retry` - this the key difference to the HedgeHog implementation which would throw away the whole `Gen`.